### DISCUSSION TITLE
after update to 4.9.0-0.okd-2022-01-29-035536 `oc` sometimes returns stale data

### DISCUSSION THREAD

**Discussion by ibotty** [okd-project/okd#1117] [General]
**Describe the bug**
Many things in the cluster behave strangely.  I have many `CrashLoopBackoffs` in `openshift` namespaces and not pods not starting.  I assume the reason is the following phenomenon.

When getting any data via the apiserver (e.g. via `oc`)  I **sometimes** get stale results.

Example:
```
❯ oc get node
NAME                                  STATUS   ROLES           AGE   VERSION
compute1.cluster.<redacted>   Ready    worker          71d   v1.22.3+2cb6068
compute2.cluster.<redacted>   Ready    worker          71d   v1.22.3+2cb6068
control1.cluster.<redacted>   Ready    master,worker   84d   v1.22.3+2cb6068
control2.cluster.<redacted>   Ready    master,worker   84d   v1.22.3+2cb6068
control3.cluster.<redacted>   Ready    master,worker   84d   v1.22.3+2cb6068
❯ oc get node
NAME                                  STATUS                     ROLES           AGE   VERSION
compute1.cluster.<redacted>   Ready                      worker          71d   v1.22.3+2cb6068
compute2.cluster.<redacted>   Ready                      worker          71d   v1.22.3+2cb6068
control1.cluster.<redacted>   Ready,SchedulingDisabled   master          84d   v1.22.3+2cb6068
control2.cluster.<redacted>   Ready,SchedulingDisabled   master,worker   84d   v1.22.3+2cb6068
control3.cluster.<redacted>   Ready                      master,worker   84d   v1.22.3+2cb6068
```

When watching for changes, I immediatly see whether the data is stale, because I run into a `too old resource version`:

```
❯ oc get node -w
NAME                                  STATUS                     ROLES           AGE   VERSION
compute1.cluster.<redacted>   Ready                      worker          71d   v1.22.3+2cb6068
compute2.cluster.<redacted>   Ready                      worker          71d   v1.22.3+2cb6068
control1.cluster.<redacted>   Ready,SchedulingDisabled   master          84d   v1.22.3+2cb6068
control2.cluster.<redacted>   Ready,SchedulingDisabled   master,worker   84d   v1.22.3+2cb6068
control3.cluster.<redacted>   Ready                      master,worker   84d   v1.22.3+2cb6068
STATUS                                REASON                     MESSAGE
Failure                               Expired                    too old resource version: 54106678 (54439732)
```

This also happens when only speaking to specific nodes:
```
❯ oc --tls-server-name api.cluster.<redacted> -s https://control2.cluster.<redacted>:6443 get node  -w
NAME                                  STATUS   ROLES           AGE   VERSION
compute1.cluster.<redacted>   Ready    worker          71d   v1.22.3+2cb6068
compute2.cluster.<redacted>   Ready    worker          71d   v1.22.3+2cb6068
control1.cluster.<redacted>   Ready    master,worker   84d   v1.22.3+2cb6068
control2.cluster.<redacted>   Ready    master,worker   84d   v1.22.3+2cb6068
control3.cluster.<redacted>   Ready    master,worker   84d   v1.22.3+2cb6068
❯ oc --tls-server-name api.cluster.<redacted> -s https://control2.cluster.<redacted>:6443 get node -w
NAME                                  STATUS                     ROLES           AGE   VERSION
compute1.cluster.<redacted>   Ready                      worker          71d   v1.22.3+2cb6068
compute2.cluster.<redacted>   Ready                      worker          71d   v1.22.3+2cb6068
control1.cluster.<redacted>   Ready,SchedulingDisabled   master          84d   v1.22.3+2cb6068
control2.cluster.<redacted>   Ready,SchedulingDisabled   master,worker   84d   v1.22.3+2cb6068
control3.cluster.<redacted>   Ready                      master,worker   84d   v1.22.3+2cb6068
STATUS                                REASON                     MESSAGE
Failure                               Expired                    too old resource version: 54107742 (54442257)
```

(The same happens on the other control plane nodes.)

The `too old resource version` keeps increasing.

**Version**
4.9.0-0.okd-2022-01-29-035536

**Log bundle**

**Comment by vrutkovs** [comment]
Please attach (or upload to the public file sharing service) [must-gather](https://docs.okd.io/latest/support/gathering-cluster-data.html#support_gathering_data_gathering-cluster-data) archive

**Reply by ibotty** [reply]
https://fedorapeople.org/~ibotty/must-gather-2022-02-22.tar.gz with sha256sum `62fde64c9d072f71c326f0382776f4a750acfd3bbc53da158ec3acc140b4cfdc`.

**Reply by vrutkovs** [reply]
`clusterversion` says its still progressing:
>Working towards 4.9.0-0.okd-2022-01-29-035536: 97 of 745 done (13% complete)

Two odd pieces in this must gather though:

* no logs in `openshift-cluster-version` namespace. Did `must-gather` throw any errors? CVO logs would help us check which manifests is it looping at

* MCO operator can't update the daemonset:
>Unable to apply 4.9.0-0.okd-2022-01-29-035536: Operation cannot be fulfilled
      on daemonsets.apps "machine-config-daemon": the object has been modified; please
      apply your changes to the latest version and try again

Not sure why, try removing `ds/machine-config-daemon` in `openshift-machine-config-operator` namespace?

**Reply by ibotty** [reply]
There are many Problems with the `must-gather`, because I guess it cannot schedule many pods.  Anything of particular interest I can gather manually?

When deleting the `ds/machine-config-daemon` it does not get recreated.  Interestingly (I include `-w`, so this is up-to-date):
```
oc get mcp -w
NAME     CONFIG                                             UPDATED   UPDATING   DEGRADED   MACHINECOUNT   READYMACHINECOUNT   UPDATEDMACHINECOUNT   DEGRADEDMACHINECOUNT   AGE
master   rendered-master-853061bbec3c27fc3781a325716a5902   True      False      False      3              3                   3                     0                      84d
worker   rendered-worker-9f71cf4296898a76560deb108b55ffa3   True      False      False      2              2                   2                     0                      84d
```
Which corresponds to the current (new) `mc`: 
```
rendered-master-207e8053d9c3d93fde0beb95db3b89f0   178194ec84ff3aa3978279d9f45307f549cd1d8f   3.2.0             84d
rendered-master-3ccf6c0867f25f609b9b902202df87d9   03a31c8d9e27b858181dcf857b1e53800856553a   3.2.0             72d
rendered-master-853061bbec3c27fc3781a325716a5902   2909118923c2e01690115e9bf27c7aba779dd01c   3.2.0             23h
rendered-master-c1c57860df829f32d4887e4f514b6a18   2909118923c2e01690115e9bf27c7aba779dd01c   3.2.0             36d
rendered-worker-0a7f9568bc0fffa32d2d01c1bb7c30ca   03a31c8d9e27b858181dcf857b1e53800856553a   3.2.0             72d
rendered-worker-198e7e9f69ae09d235cbe20fbfd31632   178194ec84ff3aa3978279d9f45307f549cd1d8f   3.2.0             84d
rendered-worker-39c991bf95b3a7e2c5d22bcebb9b4a9d   2909118923c2e01690115e9bf27c7aba779dd01c   3.2.0             36d
rendered-worker-9f71cf4296898a76560deb108b55ffa3   2909118923c2e01690115e9bf27c7aba779dd01c   3.2.0             23h
```

**Reply by ibotty** [reply]
Argh.  Now github ate my reply.  In short:

```
❯ oc get po -n openshift-kube-apiserver -w
NAME                                                     READY   STATUS             RESTARTS        AGE
kube-apiserver-control1.cluster.<redacted>       3/5     CrashLoopBackOff   201 (99s ago)   4h2m
kube-apiserver-control2.cluster.<redacted>       5/5     Running            58 (28m ago)    23h
kube-apiserver-control3.cluster.<redacted>       5/5     Running            84 (138m ago)   23h
revision-pruner-32-control1.cluster.<redacted>   0/1     Completed          0               16h
revision-pruner-32-control2.cluster.<redacted>   0/1     Completed          0               6h56m
revision-pruner-32-control3.cluster.<redacted>   0/1     Completed          0               5h42m
```

The logs of the failing `kube-apiserver` can be found at https://fedorapeople.org/~ibotty/openshift-kube-apiserver-2022-02-22.log.

**Reply by vrutkovs** [reply]
>it cannot schedule many pods

That's weird, the cluster looks healthy, all nodes are running. Are they overprovisioned? Another weird thing I noticed - masters are labelled as workers, so this might put them into additional stress. With dedicated workers this is usually not advised. Try removing non-control-plane components for upgrade to complete?

I think last 50 lines of `openshift-cluster-version-operator` pod would be most interesting to figure out current state.

>When deleting the ds/machine-config-daemon it does not get recreated

That's okay, the operator would catch up eventually. If not - nuke `machine-config-operator` pod too. It may sound scary, but generally you need those to apply MC changes so be aware of that

**Reply by ibotty** [reply]
I uploaded `oc logs -n openshift-cluster-version cluster-version-operator-5b857b666-g2p5l` to https://fedorapeople.org/~ibotty/cluster-version-operator-5b857b666-g2p5l.log.

**Reply by vrutkovs** [reply]
Right, now it shows that kube-api is degraded. From the failing log:
```
2022-02-22T14:42:44.097090050Z F0222 14:42:44.096986      20 controller.go:161] Unable to perform initial IP allocation check: Operation cannot be fulfilled on serviceipallocations "": the provided resource version does not match
2022-02-22T14:42:44.218112491Z goroutine 4026 [running]:
2022-02-22T14:42:44.218112491Z k8s.io/kubernetes/vendor/k8s.io/klog/v2.stacks(0xc000010001, 0xc0013bcc30, 0xc7, 0xc9)
2022-02-22T14:42:44.218112491Z 	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1026 +0xb9
2022-02-22T14:42:44.218112491Z k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).output(0x87b17a0, 0xc000000003, 0x0, 0x0, 0xc000a9e230, 0x0, 0x7094558, 0xd, 0xa1, 0x0)
2022-02-22T14:42:44.218112491Z 	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:975 +0x1e5
2022-02-22T14:42:44.218112491Z k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).printf(0x87b17a0, 0xc000000003, 0x0, 0x0, 0x0, 0x0, 0x56c646d, 0x31, 0xc00fcf4ec0, 0x1, ...)
2022-02-22T14:42:44.218112491Z 	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:753 +0x19a
```

That looks like a genuine bug, please file one for kube-apiserver - https://bugzilla.redhat.com/enter_bug.cgi?product=OpenShift%20Container%20Platform&component=kube-apiserver&version=4.9 and attach the must-gather to this issue.

Try rebooting the control1 node as well, this might bring kapi up to speed?

**Reply by ibotty** [reply]
> > it cannot schedule many pods
> 
> That's weird, the cluster looks healthy, all nodes are running. Are they overprovisioned? Another weird thing I noticed - masters are labelled as workers, so this might put them into additional stress. With dedicated workers this is usually not advised. Try removing non-control-plane components for upgrade to complete?

I don't think that's the case.

```
[core@control1 ~]$ free -h
               total        used        free      shared  buff/cache   available
Mem:            15Gi       4.8Gi       3.1Gi       208Mi       7.3Gi       9.9Gi
Swap:             0B          0B          0B
```

The two worker nodes are running a DB that sees some heavy usage, the rest of the cluster is not doing much.

Now that I restarted the `control1` node, `kube-apiserver` is running on that node again.  It still sometimes returns stale data (as described above) though.

The `openshift-machine-config-operator` also crashes often.  The logs can be found at https://fedorapeople.org/~ibotty/machine-config-operator-6754676c8d-5fss9.log.

**Reply by vrutkovs** [reply]
> It still sometimes returns stale data (as described above) though.

This might be the result of DB pods on master :( Could you monitor for ~hour and see if it recovers?

MCD pods are crashing with:
>failed to renew lease openshift-machine-config-operator/machine-config: timed out waiting for the condition

which is expected, as the API is not fully up

**Reply by ibotty** [reply]
Which API is not up?   The `kube-apiserver` pods are `Running` since the reboot of that node.  Anything else I can debug?

**Comment by ibotty** [comment]
Oh, the DBs are not used now that the cluster is not running and they are running **not** on the masters.
